### PR TITLE
Adds GS Distribution Test and Tracer

### PR DESCRIPTION
**Pull request by nils-wisiol**

### CODE REVIEW

File: victims/GS/tracer.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+class Tracer(connections.Tracer):
+
+    @classmethod
+    def random_seed(cls, seed: str, length: int) -> List[int]:
+        return list(cls.rng(seed).integers(low=0, high=2 ** 8 - 1, size=length))
+
+    def trigger(self, seed: str, poly_length: int = 12) -> Tuple:
+        seed = self.random_seed(f'GS {seed}', 16) 
+        samples_length = 2
```
Comment (nils-wisiol) [inline victims/GS/tracer.py:18]: Let's do one sample at a time, so we can understand the individual traces easier

File: victims/GS/test_GS.py [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
+        victim.receive_return_code()
+
+        # Send the random seed
+        victim.send_array('k', 'H', k)
+
+        # Convert the received values to signed int16 values
+        results = victim.receive_array('q', little=True)[1]
+
+        # Check if the statistic of result are correct
+        mu, sigma = norm.fit(results)  # noqa
```
Comment (nils-wisiol) [inline victims/GS/test_GS.py:22]: We need to assert somehow that the distribution approximately fits a Gaussian distribution

File: victims/GS/test_GS.py [hunk 3]
```diff
@@ -1,15 +1,22 @@
...
-    )
+from numpy.random import default_rng
+from scipy.stats import norm
+
+
+def test_samples_correctness(victim):
+    samples_length = [512]
+    for i in range(1):
+        # Create a random key
+        k = list(default_rng(256).integers(low=0, high=2**8 - 1, size=16))
```
Comment (nils-wisiol) [inline victims/GS/test_GS.py:9]: The `256` here is a bit confusing: it's the seed for the PRNG, it has nothing to do with the output size. Any number would do, why choose this one?

**Comment by nils-wisiol** [comment]
We probably need to select the tracing time more carefully: currently, the samples takes longer to generate two samples than we can trace. (Max trace length is about 24k.)

Even after so many samples, the process apparently didn't finish:

image

File: victims/GS/main.c [hunk 4]
```diff
@@ -1,96 +1,125 @@
...
+        result[i] = 0;
     }
 
     /* Start of calculations */
-     status = randomness_set_seed(key);
-     trigger_high();
-     status = gaussian_sampler(result,result_length);
-     trigger_low();
+    status = randomness_set_seed(key);
+    trigger_high();
```
Comment (nils-wisiol) [inline victims/GS/main.c:44]: Move this to after AES

**Comment by Soundes-M** [comment]
> We probably need to select the tracing time more carefully: currently, the samples takes longer to generate two samples than we can trace. (Max trace length is about 24k.)
> 
> Even after so many samples, the process apparently didn't finish:
> 
> image

I moved the trigger after the randomness_generator. I also tried to "optimize" the code and use smaller variables without affecting the accuracy of the results. (see the ode I reduced from uint64_t to uint8_t). That did not change the traces' length.  
I had an idea to trace only the rejection condition (see the last commit) instead of (sampling+rejection). However I am not sure if we can predict the sample by doing that.

**Comment by nils-wisiol** [comment]
>  (see the last commit)

I think you didn't add any new commits?

**Comment by Soundes-M** [comment]
> > (see the last commit)
> 
> I think you didn't add any new commits?
on TestGS branch. DId you move the code to TestGS2?

**Comment by nils-wisiol** [comment]
Yeah [I took the code from TestGS,](https://github.com/nils-wisiol/m4sc/pull/17#issuecomment-659949109) rebased it on top of master, cleaned it up, removed some merge errors and put it in TestGS2.

File: victims/GS/test_GS.py [hunk 5]
```diff
@@ -1,22 +1,33 @@
...
+import numpy as np
+import pytest
+ 
 def test_samples_correctness(victim):
-    samples_length = [512]
-    for i in range(1):
-        # Create a random key
-        k = list(default_rng(256).integers(low=0, high=2**8 - 1, size=16))
-
+    samples_length = [1024]
```
Comment (nils-wisiol) [inline victims/GS/test_GS.py:13]: I believe this won't work on chipwhisperer

File: victims/GS/main.c [hunk 6]
```diff
@@ -38,19 +38,56 @@ static uint8_t recv_key(int size, uint16_t *data) {
...
-    trigger_high();
-    status = gaussian_sampler(result, result_length);
-    trigger_low();
+     status = randomness_set_seed(key);
+     
+     while (j < result_length)
+     {  
+         
+        
+	do
```
Comment (nils-wisiol) [inline victims/GS/main.c:57]: adjust formatting in the following lines :)

File: victims/GS/test_GS.py [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
+ 
+def test_samples_correctness(victim):
+    samples_length = [1024]
+    for i in range(5):
+        # Send the length of samples
+        victim.send_array('s', 'H', samples_length)
+        victim.receive_return_code()
+
+        # Create a random key
+        k = list(default_rng().integers(low=0, high=2 ** 16 - 1, size=16))
```
Comment (nils-wisiol) [inline victims/GS/test_GS.py:20]: add a seed argument to `default_rng` to obtain reproducible results

**Pull request closed without merging**